core/domain/exp_services.py

```diff
@@ -513,27 +507,13 @@ def get_exploration_ids_matching_query(query_string, cursor=None):
         exp_ids, search_cursor = search_services.search_explorations(
             query_string, remaining_to_fetch, cursor=search_cursor)
 
-        invalid_exp_ids = []
-        for ind, model in enumerate(
+        for ind, _ in enumerate(
                 exp_models.ExpSummaryModel.get_multi(exp_ids)):
-            if model is not None:
-                returned_exploration_ids.append(exp_ids[ind])
```

Comment:
Ditto, can you please explain why this method is changed?

Reply:
Since we are making a query for explorations, we won't get an exploration that does not has an associated model. You can check the search function in gae_search_services. It searches through models

Reply:
I think we are assuming here that all the serach_indexes will always have a corresp[onding ExplorationModel. Let's not assume that.

Reply:
I think the search_explorations function only fetch the explorations which has an exp model. It doesn't fetch empty Exploration domain objects, I tried it

Reply:
Search services fetch the search indexed documents of exploration namespace (which isn't the exploration model), we create such documents out of exp_models to provide full-text search functionality.

Assume a scenario where we have a search indexed document of exploration for which the exp_model is deleted or marked unpublished.

Reply:
I tried both unpublishing but nothing works. I think this is because search_cursor is still None?

Reply:
> I tried both unpublishing but nothing works.

I'm unable to understand this.

Overall,a scenario where there exists a search_document for and exploration but the exploration is deleted. In that case model will be none and we expect this method to raise exception.

Reply:
I meant I tried deleting a model by `delete_exploration` and also tried unpublishing an exploration but still those lines are not getting hit.
> where there exists a search_document for and exploration

Can you tell me how can I keep the search document but delete the model? Because whenever I am trying to delete the model or unpublish the exploration, the search document also gets removed.

Reply:
I think we are rasing exception here for the worst cases which can be raised because of two directly connected services, i.e, while saving exploration we are saving ExplorationModel as well as Creating searchIndex there is a chance of one to happen and one to fail because of some unknown reasons and if we won't check for such exception and raise a log then we won't be able to know why things are failing. I think you should do git blame on this line to check when and why this was implemented (For more details.) As well as read the docstring for this method.

> Can you tell me how can I keep the search document but delete the model?

For writing tests, you can create exploration and explicitly delete the search index using servach_services.

Reply:
You'll find [this commit](https://github.com/oppia/oppia/commit/76f624dba22274dafa916263936afbc0e998c740#diff-6abb716589c61217f40bf3386182859cL234) useful.

Reply:
I mocked `delete_documents_from_index` to do nothing. That way the exploration does not get deleted from the document.